Add ML tutorials (#7180)
opensearch-project/documentation-website: _ml-commons-plugin/tutorials/build-chatbot.md

{: .warning}

Sometimes a large language model (LLM) cannot answer a question right away. For example, an LLM can't tell you how many errors there are in your log index for last week because its knowledge base does not contain your proprietary data. In this case, you need to provide additional information to an LLM in a subsequent call. You can use an agent to solve such complex problems. The agent can run tools to obtain more information from configured data sources and send the additional information to the LLM as context.

This tutorial describes how to build your own chatbot in OpenSearch using a `conversational` agent. For more information about agents, see [Agents and tools]({{site.url}}{{site.baseurl}}/ml-commons-plugin/agents-tools/index/).

Replace the placeholders starting with the prefix `your_` with your own values.
{: .note}

## Prerequisite

Log in to the OpenSearch Dashboards homepage, select **Add sample data**, and add **Sample eCommerce orders** data.

## Step 1: Configure a knowledge base

> Follow Prerequisite and Step 1 of the [RAG with a conversational flow agent tutorial]({{site.url}}{{site.baseurl}}/ml-commons-plugin/tutorials/rag-conversational-agent/) to configure the `test_population_data` knowledge base index, which contains US city population data.

Author: ```suggestion
Meet the prerequisite and follow Step 1 of the [RAG with a conversational flow agent tutorial]({{site.url}}{{site.baseurl}}/ml-commons-plugin/tutorials/rag-conversational-agent/) to configure the `test_population_data` knowledge base index, which contains US city population data.
```

Next, create and test an agent.

### Create an agent

Create an agent of the `conversational` type. 

The agent is configured with the following information:

- Meta information: `name`, `type`, `description`.
- LLM information: The agent uses an LLM to reason and select the next step, including choosing an appropriate tool and preparing the tool input.
- Tools: A tool is a function that can be executed by the agent. Each tool can define its own `name`, `description`, and `parameters`.
- Memory: Stores chat messages. Currently, OpenSearch only supports one memory type: `conversation_index`.

The agent contains the following parameters:

- `conversational`: This agent type has a built-in prompt. To override it with your own prompt, see [Step 4](#step-4-optional-create-an-agent-with-a-custom-prompt).
- `app_type`: Specify this parameter for reference purposes in order to differentiate between multiple agents.
- `llm`: Defines the LLM configuration:
   - `"max_iteration": 5`:  The agent runs the LLM a maximum of five times.
   - `"response_filter": "$.completion"`: Needed to retrieve the LLM answer from the Bedrock Claude model response.

>    - `"message_history_limit": 5`: The agent retrieves a maximum of the five most recent history messages and adds them to the LLM context. Set this parameter to `0` to omit message history in the context.

Author: ```suggestion
   - `"message_history_limit": 5`: The agent retrieves a maximum of the five most recent historical messages and adds them to the LLM context. Set this parameter to `0` to omit message history in the context.
```

>    - `"message_history_limit": 5`: The agent retrieves a maximum of the five most recent history messages and adds them to the LLM context. Set this parameter to `0` to omit message history in the context.

Author: ```suggestion
   - `"message_history_limit": 5`: The agent retrieves a maximum of the five most recent history messages and adds them to the LLM context. Set this parameter to `0` to omit message history in the context.
```
```suggestion
   - `"message_history_limit": 5`: The agent retrieves a maximum of the five most recent historical messages and adds them to the LLM context. Set this parameter to `0` to omit message history in the context.
```

   - `disable_trace`: If `true`, the agent does not store trace data in memory. Trace data is included in each message and provides a detailed recount of steps performed while generating the message.
- `memory`: Defines how to store messages. Currently, OpenSearch only supports the `conversation_index` memory, which stores messages in a memory index.
- Tools: 
   - An LLM will reason to decide which tool to run and will prepare the tool's input. 
   - To include the tool's output in the response, specify `"include_output_in_agent_response": true`. In this tutorial, you will include the `PPLTool` output in the response (see the example response in [Test the agent](#test-the-agent)). 
   - By default, the tool's `name` is the same as the tool's `type`, and each tool has a default description. You can override the tool's `name` and `description`.

>    - Each tool in the `tools` list must have a unique name. For example, the following demo agent defines two tools of the `VectorDBTool` type with different names (`population_data_knowledge_base` and `stock_price_data_knowledge_base`). Each tool has a custom description so the LLM can easily understand what the tool does.

Author: ```suggestion
   - Each tool in the `tools` list must have a unique name. For example, the following demo agent defines two tools of the `VectorDBTool` type with different names (`population_data_knowledge_base` and `stock_price_data_knowledge_base`). Each tool has a custom description so that the LLM can easily understand what the tool does.
```

    },
    {
      "type": "SearchAnomalyDetectorsTool"
    },
    {
      "type": "SearchAnomalyResultsTool"
    },
    {
      "type": "SearchMonitorsTool"
    },
    {
      "type": "SearchAlertsTool"
    }
  ]
}
```
{% include copy-curl.html %}

Note the agent ID; you'll use it in the next step.

### Test the agent

Note the following testing tips: 

- You can view the detailed steps of an agent execution in one of the following ways:
   - Enable verbose mode: `"verbose": true`.
   - Call the Get Trace API: `GET _plugins/_ml/memory/message/your_message_id/traces`.

- An LLM may hallucinate. It may choose a wrong tool to solve your problem, especially when you have configured many tools. To avoid hallucinations, try the following options:
   - Avoid configuring many tools in an agent.
   - Provide a detailed tool description to clarify what the tool can do. 
   - Specify the tool to use in the LLM question, for example, `Can you use the PPLTool to query the opensearch_dashboards_sample_data_ecommerce index so it can calculate how many orders were placed last week?`.

>    - Specify the tool to use when executing an agent. For example, specify to only use the `PPLTool` and `CatIndexTool` to process the current request.

Author: ```suggestion
   - Specify the tool to use when executing an agent. For example, specify that only `PPLTool` and `CatIndexTool` should be used to process the current request.
```

"""
        },
        {
          "name": "response",
          "result": "According to the population data tool, the population of Seattle increased by approximately 28,000 people from 2021 to 2023, which is a 0.82% increase from 2021 to 2022 and a 0.86% increase from 2022 to 2023."
        }
      ]
    }
  ]
}
```

Obtain trace data:

```json
GET _plugins/_ml/memory/message/L0JuyI0Bn3OCesyv3-Er/traces
```
{% include copy-curl.html %}

#### Test conversational memory

To continue the same conversation, specify the conversation's `memory_id` when executing the agent:

```json
POST _plugins/_ml/agents/your_agent_id/_execute
{
  "parameters": {
    "question": "What's the population of Austin 2023, compare with Seattle",
    "memory_id": "LkJuyI0Bn3OCesyv3-Ef",
    "verbose": true
  }
}
```
{% include copy-curl.html %}

> In the response, note that the `population_data_knowledge_base` doesn't return the population of Seattle. Instead, the agent learns the population of Seattle from historical messages:

Author: ```suggestion
In the response, note that the `population_data_knowledge_base` doesn't return the population of Seattle. Instead, the agent learns the population of Seattle by referencing historical messages:
```

"""
        },
        {
          "name": "response",
          "result": "According to the population data tool, the population of Austin in 2023 is approximately 2,228,000 people, a 2.39% increase from 2022. This is lower than the population of Seattle in 2023 which is approximately 3,519,000 people, a 0.86% increase from 2022."
        }
      ]
    }
  ]
}
```

View all messages:

```json
GET _plugins/_ml/memory/LkJuyI0Bn3OCesyv3-Ef/messages
```
{% include copy-curl.html %}

Obtain trace data:

```json
GET _plugins/_ml/memory/message/00J6yI0Bn3OCesyvIuGZ/traces
```
{% include copy-curl.html %}

## Step 4 (Optional): Create an agent with a custom prompt 

All agents have the following default prompt:

```json
"prompt": """

Human:${parameters.prompt.prefix}

${parameters.prompt.suffix}

Human: follow RESPONSE FORMAT INSTRUCTIONS

Assistant:"""
```

The prompt consists of two parts:

> - `${parameters.prompt.prefix}`: A prompt prefix that describes what the AI Assistant can do. You can change this parameter based on your use case, for example, `You are a professional data analyst. You will always answer questions based on the tool response first. If you don't know the answer, just say you don't know.`

Author: ```suggestion
- `${parameters.prompt.prefix}`: A prompt prefix that describes what the AI assistant can do. You can change this parameter based on your use case, for example, `You are a professional data analyst. You will always answer questions based on the tool response first. If you don't know the answer, just say you don't know.`
```

Assistant should NEVER suggest run a tool with same input if it's already in TOOL RESPONSE. 
The tools the human can use are:

${parameters.tool_descriptions}

${parameters.chat_history}

${parameters.prompt.format_instruction}

Human:USER'S INPUT
--------------------
Here is the user's input :
${parameters.question}

${parameters.scratchpad}"""
```

The `prompt.suffix` consists of the following placeholders:

- `${parameters.tool_descriptions}`: This placeholder will be filled with the agent's tool information: the tool name and description. If you omit this placeholder, the agent will not use any tools.
- `${parameters.prompt.format_instruction}`: This placeholder defines the LLM response format. This placeholder is critical and we recommend not removing it.
- `${parameters.chat_history}`: This placeholder will be filled with the message history of the current memory. If you don't set the `memory_id` when you run the agent, or there are no history messages, this placeholder will be empty. If you don't need chat history, you can remove this placeholder.
- `${parameters.question}`: This placeholder will be filled with the user question.

> - `${parameters.scratchpad}`: This placeholder will be filled with the detailed steps in agent execution. These steps are the same as those you can view by specifying verbose mode or obtaining trace data (see an example in [Test the agent](#test-the-agent)). This placeholder is critical for the LLM to reason and select the next step based on the outcome of the previous steps. We recommend not removing this placeholder.

Author: No, just referring to the mode here

      },
      "include_output_in_agent_response": true
    }
  ],
  "memory": {
    "type": "conversation_index"
  }
}
```
{% include copy-curl.html %}

> Note the root chatbot agent ID, then log in to your OpenSearch server, go to the OpenSearch config folder (`$OS_HOME/config`), and run the following command:

Author: Same here. Generic concept
